content/docs/ai/ai-langchain.md

```diff
@@ -0,0 +1,384 @@
+  SQLQuery: "SELECT name, age FROM public.passenger WHERE age IS NOT NULL ORDER BY age DESC LIMIT 1;"
+"""
+
+# Define Custom Prompt
+TEMPLATE = """Given an input question, first create a syntactically correct {dialect} query to run, then look at the results of the query and return the answer.
+Use the following format:
+
+Question: "Question here"
+SQLQuery: "SQL Query to run"
+SQLResult: "Result of the SQLQuery"
+Answer: "Final answer here"
+
+Only use the following tables:
+
+{table_info}
+
+Some examples of SQL queries that correspond to questions are:
+
+{few_shot_examples}
+
+Question: {input}"""
+
+CUSTOM_PROMPT = PromptTemplate(
+    input_variables=["input", "few_shot_examples", "table_info", "dialect"], template=TEMPLATE
+)
+
+# Setup the database chain
+db_chain = SQLDatabaseChain(llm=llm, database=db, verbose=True)
+
+def get_prompt():
+    print("Type 'exit' to quit")
+    while True:
+        prompt = input("Ask a question or type exit to quit: ")
+
+        if prompt.lower() == 'exit':
+            print('Exiting...')
+            break
+        else:
+            try:
+                question = CUSTOM_PROMPT.format(
+                    input=prompt,
+                    few_shot_examples=few_shot_examples,
+                    table_info=table_info,
+                    dialect="PostgreSQL"
+                )
+                print(db_chain.run(question))
+            except Exception as e:
+                print(e)
+
+get_prompt()
+```
+
+<details>
+<summary>Understanding the app.py code</summary>
+
+- **Importing Necessary Modules**:
+  - `os`: Provides a portable way of using operating system-dependent functionality, mainly for accessing environment variables.
+  - `psycopg2`: The PostgreSQL adapter for Python, allowing interaction with the PostgreSQL database.
+  - `environ`: Used to manage environment variables and `.env` files.
+  - `urlparse`: Part of Python's standard library for parsing URLs.
+  - `langchain` related imports: A suite of utilities and modules for querying the database using natural language.
+
+- **Setup and Read the .env File**:
+  - The `environ` module is used to read environment variables, specifically the database connection details, from a `.env` file.
+
+- **Extract Connection Details from the .env File**:
+  - The connection string is retrieved from the `.env` file and parsed to extract database connection details like username, password, host, port, and database name.
+
+- **Setup Database Connection**:
+  - The `SQLDatabase.from_uri()` method from the `langchain.utilities` module establishes a connection to the PostgreSQL database using the extracted details.
+
+- **Setup Language Learning Model (llm)**:
+  - An instance of `OpenAI` from `langchain.llms` is created, with certain parameters like temperature. It also requires an API key, which is fetched from the environment variables.
+
+- **Defining Table Information and Few Shot Examples**:
```

Comment:
Should the term be hyphenated: Few-Shot 
?

Reply:
This one is OK. This is how it's defined in the OpenAI docs. It's their term: https://python.langchain.com/docs/use_cases/qa_structured/sql